### ISSUE TITLE
Camera Orientation to _locator is Incorrect

### ISSUE THREAD

**Issue by acbuynak** [osu-capstone-afrl/photogrammetry-capture-motion#2]
Issue where the Z-Axis does not properly orient to point at the central `_locator`

Using method's highlighted here: https://math.stackexchange.com/questions/180418/calculate-rotation-matrix-to-align-vector-a-to-vector-b-in-3d/476311#476311

![image](https://user-images.githubusercontent.com/44533530/109197590-30458300-776b-11eb-88e0-0b1f4368082b.png)

**Comment by acbuynak** [comment]
#### Description of Methods
Methods 2&3 picked from https://math.stackexchange.com/questions/180418/

**Method 2**
![image](https://user-images.githubusercontent.com/44533530/109230762-626adb00-7793-11eb-80ff-1d0fb3740116.png)

This method is applied here:
https://github.com/osu-capstone-afrl/photogrammetry-capture-motion/blob/5a08124ee6d1cd76ff96013f0527d9e353949ba1/scripts/path_plans.py#L277-L283

**Method 3**
![image](https://user-images.githubusercontent.com/44533530/109230281-9bef1680-7792-11eb-8b05-08b214ee3869.png)

Applied in code here:
https://github.com/osu-capstone-afrl/photogrammetry-capture-motion/blob/5a08124ee6d1cd76ff96013f0527d9e353949ba1/scripts/path_plans.py#L290-L302

**Comment by acbuynak** [comment]
Method Update 2

**Comment by acbuynak** [comment]
I manually iterated through different angles and observed (below) how the resultant rotation was reflected in the first point.

### Setup
Only two frames were plotted. The fixed frame at the origin (shown on the right) and a single camera pose angle located along the `- x` axis. Using such a setup ensures the rotation is about the yellow `y` axis of the body frame. The angle rotated may be determined by backing out `theta` from the resultant rotation matrix `rot_matrix` in the code.

![image](https://user-images.githubusercontent.com/44533530/109226343-9989be00-778c-11eb-9958-3338e7709ce1.png)

For a rotation about the `y-axis`, the rotation matrix should resemble the following...
![image](https://user-images.githubusercontent.com/44533530/109226650-20d73180-778d-11eb-83a2-6ab0ea950574.png)



### **Results Table**
| Input Angle (deg)(positive) | Result | | Input Angle (deg)(negative) | Result |
|------------------------------|----------|-|-------------------------------|----------|
| 180 <br>[[ 0. -0. -1.]<br> [-0.  1. -0.]<br>[ 1.  0.  0.]]| n/a  |=| -180 |  n/a  |
| 170 | ![image](https://user-images.githubusercontent.com/44533530/109223195-4a418e80-7788-11eb-840b-9f4db66eaf46.png)  |=| -170 |  ![image](https://user-images.githubusercontent.com/44533530/109223222-5299c980-7788-11eb-8a97-17ca1cc37ab2.png)  |
| 165| ![image](https://user-images.githubusercontent.com/44533530/109226091-326c0980-778c-11eb-8a61-46aeefb1d0f4.png)|=|-165| ![image](https://user-images.githubusercontent.com/44533530/109226196-5891a980-778c-11eb-8172-fbfff4a81697.png) |
| 160 |![image](https://user-images.githubusercontent.com/44533530/109223289-647b6c80-7788-11eb-96e1-27bead9be034.png)|=| -160 |![image](https://user-images.githubusercontent.com/44533530/109223311-6cd3a780-7788-11eb-8e71-588367a2c9c5.png)    |
| 155 |  ![image](https://user-images.githubusercontent.com/44533530/109223321-71985b80-7788-11eb-9e38-256a35fe66c5.png) |=| -155 | ![image](https://user-images.githubusercontent.com/44533530/109223336-765d0f80-7788-11eb-850a-090c7bb0e862.png)  |
| 100 | ![image](https://user-images.githubusercontent.com/44533530/109223356-7bba5a00-7788-11eb-8eaa-6ce808895b46.png)  |=| -100 |  ![image](https://user-images.githubusercontent.com/44533530/109223371-7fe67780-7788-11eb-800b-be49c89d84f8.png)  |
| 90 |  ![image](https://user-images.githubusercontent.com/44533530/109225466-4b27ef80-778b-11eb-8df8-de693ac18b8a.png) |=| -90 |  ![image](https://user-images.githubusercontent.com/44533530/109225484-50853a00-778b-11eb-8e33-378dc7024ed5.png)|
| 80 | ![image](https://user-images.githubusercontent.com/44533530/109225494-5549ee00-778b-11eb-9071-c420fbc26e94.png)  |=| -80 | ![image](https://user-images.githubusercontent.com/44533530/109225510-5aa73880-778b-11eb-8e5f-1159aac5610e.png) |
| 45 |![image](https://user-images.githubusercontent.com/44533530/109225524-609d1980-778b-11eb-850e-7688e60745bd.png)   |=| -45 | ![image](https://user-images.githubusercontent.com/44533530/109225543-65fa6400-778b-11eb-8461-d9e0342cf1bc.png) |
| 10 |  ![image](https://user-images.githubusercontent.com/44533530/109225556-6abf1800-778b-11eb-904e-d77109f56af1.png) |=| -10 | ![image](https://user-images.githubusercontent.com/44533530/109225621-82969c00-778b-11eb-85fd-7614614482ff.png)   |
| 0 | ![image](https://user-images.githubusercontent.com/44533530/109225690-9b9f4d00-778b-11eb-8483-27efec2321e7.png)  |=| 0 |    |

**Comment by acbuynak** [comment]
**Scratch all of the previous post.** No correlation should be drawn between changes in the shown code. 
After animating through all of the points in the current code design, it was realized that the resultant Rotation Matrix is varying.


Shown video increments through every rotation matrix from (-179) to (179) degrees, step size 1 degree.


https://user-images.githubusercontent.com/44533530/109242484-365a5480-77a9-11eb-9098-4991aa30bcdb.mp4

**Comment by Schellenberg3** [comment]
I'm looking at the R_y(Theta) part of the homogeneous matrix  and am noticing a few things

1.  The columns of the matrix are not unit length like they should be
2. The values of Cos(Theta) are not correct for the input angle (in degree or rad)
3. The values for Sin(theta) seem to be constant.

It might be wise to deconstruct this method (number 3 still?) further and check out where the values in each of the indexes comes from and why it deviates from the expected R_y matrix

**Comment by acbuynak** [comment]
Agreed on the need to deconstruction Method 3 to better understand the equation.

Reviewing the original post, I (re)found a comment that highlighted a potential error in the suggested solution. Said comment has not been verified by anyone else on the forum. It outputs the below motion. (see video attached)

Current Equation (ie. Original from Forum Answer):
https://github.com/osu-capstone-afrl/photogrammetry-capture-motion/blob/5a08124ee6d1cd76ff96013f0527d9e353949ba1/scripts/path_plans.py#L302

Proposed Alternative: *This matches the actual Rodriguez Formula*
```
rot_matrix = np.identity(3) + s*v_x + np.dot(v_x,v_x) * (1-c)
``` 

This seems to produce more of a looped motion, but that doesn't really tell us much.

https://user-images.githubusercontent.com/44533530/109365012-205a9b80-785e-11eb-885c-4590ecab9b1f.mp4

**Comment by Schellenberg3** [comment]
Looks better and each rotation column is unit and those values check out for cos() and sin().  Interested to see how this affect the full rotation that we want to achive!

**Comment by acbuynak** [comment]
Fixed issue in debugging visualization script where the theta (angle) being input was in degrees not radians. 
Now using a `np.deg2rad()` conversion. This produces a nice clean motion as shown below.

https://user-images.githubusercontent.com/44533530/109550154-e75c3a00-7a9c-11eb-87c8-c318a7d6be98.mp4

**Comment by acbuynak** [comment]
Digging a little deeper since this motion didn't look linear. Watching the video the middle range seems to rotate faster. Looking at the rotation matrix which found the position an issue with the rotation matrix may be found.

The below figure highlights an issue where the rotation matrix is being incorrectly generated. The theta values were derived from cells `r_11` and `r_13`.

![image](https://user-images.githubusercontent.com/44533530/109564652-5e023300-7aaf-11eb-95fa-41fa8ab3af9c.png)

**Comment by acbuynak** [comment]
#### Focusing now on a single point to help back out the issue...

Our goal is to rotation a known frame represented by unit vector `< 0, 0, 1 >` onto a locally calculated vector. The locally calculated vector represents the unit direction pointing _from_ the local point _to_ the _locator (ie part origin).

For a simple test, we will attempt to rotate the frame at camera position `( -0.25, 0, 0.25 )` to point the Z-axis at the origin `(0,0,0)`.

![image](https://user-images.githubusercontent.com/44533530/109557594-4e322100-7aa6-11eb-8469-dcc09758ffdf.png)

Or as a rotation matrix.. `rot(y,135)`
```
[[-0.707    0     0.707 ]
 [ 0        1     0     ]
 [-0.707    0    -0.707 ]]
```

As is, the equation outputs a rotation matrix of...
```
[[ 0.146  0.    -0.5   ]
 [ 0.     1.     0.    ]
 [ 0.5    0.     0.146 ]]
```
Theta derived from cos() `r_11 = 81.579 deg`
Theta derived from  sin() `r_13 = 120.000 deg`
<br>

#### Evaluating Rotation Matrix at a new point
This produces another strange dependency on position.
![image](https://user-images.githubusercontent.com/44533530/109565393-65760c00-7ab0-11eb-847d-67120bcf161f.png)

**Comment by acbuynak** [comment]
As a side observation...
Increasing the z-value of our camera location results in the following.
![image](https://user-images.githubusercontent.com/44533530/109566398-cb16c800-7ab1-11eb-9018-bdfedfeade13.png)

**Comment by acbuynak** [comment]
Resolved and removed (erroneous) dependence upon (z-value) of camera location.
Issue was the result of mistakenly using a non-unit vector to rotate about.
https://github.com/osu-capstone-afrl/photogrammetry-capture-motion/blob/739fba7ff31ae24db71e4b905993b1d37e562ce2/scripts/debug_pose_angle.py#L37

Fixed line reads: (using 2-norm instead of Frobenius)
```
v = np.cross(a,b)
v = v / np.linalg.norm(v,2)
```


#### Conclusion
This resolves any error with cells `r_11` and `r_33` as nature of cos() is independent of sign. Thus, "V" shaped result in the debugging plot like below is acceptable.

![image](https://user-images.githubusercontent.com/44533530/109578419-230afa00-7ac5-11eb-864b-df8e95e702b7.png)

Following this, it appears that the theta values derived from sin() are...
* offset by 90deg to the input
* effected by the quadrant sign

**Comment by acbuynak** [comment]
Following the fix in 322fe0e 

New Figure:
![image](https://user-images.githubusercontent.com/44533530/109683295-3792d500-7b4d-11eb-88b8-b349cac89b61.png)

**Comment by acbuynak** [comment]
Applied a new approach based on the formation of a rotation matrice based on the normalized component vectors (X, Y, Z) stacked as column vectors.

![image](https://user-images.githubusercontent.com/44533530/111041189-28d9d700-8405-11eb-8675-359134a96f9c.png)
_See Wikipedia Page [linked here.](https://en.wikipedia.org/wiki/Rotation_formalisms_in_three_dimensions#Formalism_alternatives)_

Code has not been pushed yet.

Further debugging is required. Plotting the three vectors which comprise the rotation matrix gives the LEFT plot. Applying this technique to the `path_plan` code gives the RIGHT plot. The view from both plots has been manipulated to give a similar view relative to the fixed frame.
![image](https://user-images.githubusercontent.com/44533530/111041163-e57f6880-8404-11eb-9191-56dc4ee4d3fc.png)

**Comment by acbuynak** [comment]
After doing some by-hand calculations, I believe that the rotation matrix found is correct.

The issue may lie in the actual conversion from a TF matrix to Quaternion.

### By Hand Calculations
| Point | ( x, y, z ) from Fixed Frame's perspective |
| -- | -- |
| Camera Point | `( -0.5, -0.5, -0.5 )` |
| Locator | `( 0, 0, 0 )` |
| Fixed Frame | `( 0, 0, 0 )` |

Vector from Camera Point to Locator (in Locator Frame)
`< 0.5, 0.5, -0.5 >`

New X Vector
`< -0.5, 0.5, -0.5 >`

New Y Vector
`< 0.25, 0.25, 0.5 >`

All Values Normalized
X-Axis-Vector   =  `< 0.707, -0.707,  0 >`
Y-Axis-Vector   =  `< 0.408,   0.408, 0.816 >`
Z-Axis-Vector   = `< 0.577,   0.577, -0.577 >`

Rotation Matrix ([ref](https://en.wikipedia.org/wiki/Rotation_formalisms_in_three_dimensions#Formalism_alternatives))
```
[ -0.707  0.408   0.577 ]
[  0.707  0.408   0.577 ]
[  0      0.816  -0.577 ]
```

Transformation Matrix Representation
```
[ -0.707  0.408   0.577  0 ]
[  0.707  0.408   0.577  0 ]
[  0      0.816  -0.577  0 ]
[  0      0       0      1 ]
```

Applying Translation
```
[ -0.707  0.408   0.577  -0.5 ]
[  0.707  0.408   0.577  -0.5 ]
[  0      0.816  -0.577   0.5 ]
[  0      0       0       1 ]
```

### Sanity Check
A point `(0,0,0)` in from the perspective of the Fixed Frame should equal `(0,0,||Z||)` in the perspective of the new Camera Frame.

`P_ref = T_ref,camera * P_camera`
where `P_camera = [ 0, 0, 0.866, 1 ]`
`P_ref = [ 0, 0, 0, 1 ]`

After solving, this holds true. Good!

**Comment by acbuynak** [comment]
Issue Identified. :smile: 

Having fixed the rotation matrix creation issues (using the newest axial vector-derived method), an issue where one of the tools accepted a list of XYZ rotation values and calculated the effective quanternions. The code mistakenly fed in this tool the first 3 (of 4) values of the quanternion and re-calculated the quanternion.

This has enabled the following view to be created.
![image](https://user-images.githubusercontent.com/44533530/111044229-4878fb80-8415-11eb-8901-df9e0a2c493d.png)


Further documentation of the previously attempted issues will follow in the post. Once the final code is merged into master, this issue will be closed.

**Comment by acbuynak** [comment]
## Review of Rotation Matrix Generation Methods

Each of the below presented techniques successfully rotate the Z-Axis of the camera frame to 'point' at a user-defined point.
Method 1 is excluded as they originally theory is flawed.

The below table previews the results of each (successfully) trialed rotation matrix creation method. Methods 2 & 3 produce identical results as expected. This is because they derive from same 'axis-angle' theory.

Method 4 is a more constrained technique which constrains the X-Axis' height (or Z) component to a zero value. This ensures the X-axis is constrained to a plane parallel with the fixed frames X<sub>0</sub>-Y<sub>0</sub> plane (ie. the floor).

| Method | Top View | Iso View |
| ---------- | ------------ | ----------- |
| **Method 2:** <br> Axis-Angle Direct Approach <br> [StackExchange Source](https://math.stackexchange.com/a/897677) | ![image](https://user-images.githubusercontent.com/44533530/111044533-2da78680-8417-11eb-8014-cd42bd1b10e6.png) |![image](https://user-images.githubusercontent.com/44533530/111044772-8f1c2500-8418-11eb-8e58-d32e872e1a86.png) |
| **Method 3:** <br> Axis-Angle to Rotation Matrix<br> Variant on Rodriguez Formula<br>[Wikipedia Source](https://en.wikipedia.org/wiki/Rotation_matrix#Rotation_matrix_from_axis_and_angle)<br><br>_This correctly applied the flawed answer given on [StackExchange](https://math.stackexchange.com/a/476311)_ |  ![image](https://user-images.githubusercontent.com/44533530/111044674-fab1c280-8417-11eb-9d68-b9f95fb31ac7.png) | ![image](https://user-images.githubusercontent.com/44533530/111044772-8f1c2500-8418-11eb-8e58-d32e872e1a86.png) |
| **Method 4:** <br> Axial-Vector Derived Rotation Matrix<br> _Find Z-vector, sets an X-vector orthogonal to Z-Vector on a constrained plane, uses cross-product to find Y-Vector._<br><br>_Includes various checks to avoid edge cases._  | ![image](https://user-images.githubusercontent.com/44533530/111044715-351b5f80-8418-11eb-810e-b91876055be2.png) | ![image](https://user-images.githubusercontent.com/44533530/111044736-5bd99600-8418-11eb-89bb-0f1287469a1c.png) |

**Comment by acbuynak** [comment]
Methods 2 & 3 Finalized:  87ab1e7a09ab4b7dccd14746404e4c226fe85014

Method 4 Finalized: a730cbb616e2b665cad1f81ac33a6115c43044fd

**Comment by acbuynak** [comment]
## Notes & Warning on Debug Tool

The tool created to help debug is of limited use beyond confirming visual rotations and providing animations
The figure generated to derive the angle rotated per specific elements in the rotation may mislead users.

### Debugging Tool Animation for Method 3:
Notice how the `0 deg` angle is a non-rotated translation of the Fixed Frame shown on the right.

https://user-images.githubusercontent.com/44533530/111051857-1753f780-8424-11eb-9b37-eb03795b6e28.mp4

### Debugging Tool Animation for Method 4:
Notice how the `0 deg` angle is a non-rotated translation of the Fixed Frame shown on the right.


https://user-images.githubusercontent.com/44533530/111051861-2175f600-8424-11eb-85c9-e9c01ca024d8.mp4


### Warning
The below figure was generated alongside the Method 4 animation using the debugging tool. While this does show discrepancy in angles and may mislead users to finding error in the rotation matrix, the actual rotation is acceptable as long as the rotation matrix is converted to quanternion. This will resolve any quadrant specific issues.

![image](https://user-images.githubusercontent.com/44533530/111051874-479b9600-8424-11eb-9c19-bf56a6d2ef3b.png)

**Comment by acbuynak** [comment]
Example of "Stepped Rings" method deployed with object in space.
![image](https://user-images.githubusercontent.com/44533530/111189751-ca3f6500-858c-11eb-8bbd-46d5c4cafa10.png)

**Comment by acbuynak** [comment]
Fixed rotation frame to match ROS Industrial Robot Convention for EEF TCP's.
Updated in commit; cd06dfaf38ded5ae624f856c5c3ec2452b3b3982

![image](https://user-images.githubusercontent.com/44533530/111358872-37bcc580-8661-11eb-9ece-3f628b0cedad.png)

**Comment by acbuynak** [comment]
Closing issue as rotation is correct now and meets convention.
Conversation below will direct to where documentation is kept.
